src/mlpack/tests/main_tests/nca_test.cpp

```diff
@@ -0,0 +1,380 @@
+#include <mlpack/core/optimizers/lbfgs/lbfgs.hpp>
+
+#include <boost/test/unit_test.hpp>
+#include "../test_tools.hpp"
+
+using namespace mlpack;
+
+struct NCATestFixture
+{
+ public:
+  NCATestFixture()
+  {
+    // Cache in the options for this program.
+    CLI::RestoreSettings(testName);
+  }
+
+  ~NCATestFixture()
+  {
+    // Clear the settings.
+    bindings::tests::CleanMemory();
+    CLI::ClearSettings();
+  }
+};
+
+BOOST_FIXTURE_TEST_SUITE(NCAMainTest, NCATestFixture);
+
+/**
+* Ensure that, when labels are implicitily given with input,
+* the last column is treated as labels and that we get the
+* desired shape of output.
+*/
+BOOST_AUTO_TEST_CASE(NCAExplicitImplicitLabelsTest)
+{
+  // Random dataset.
+  arma::mat x;
+  x.randu(3, 6);
+
+  SetInputParam("input", std::move(x));
+
+  mlpackMain();
+
+  // Check that last row was treated as label by checking that
+  // the output has 1 less row.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("output").n_rows, 2);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("output").n_cols, 2);
+
+  // Reset Settings.
+  CLI::ClearSettings();
+  CLI::RestoreSettings(testName);
+
+  // Now check that when labels are explicitely given, the last column
+  // of input is not treated as labels.
+  arma::mat y              = "-0.1 -0.1 -0.1  0.1  0.1  0.1;"
+                             " 1.0  0.0 -1.0  1.0  0.0 -1.0 ";
+  arma::Row<size_t> labels = " 0    0    0    1    1    1   ";
+
+  SetInputParam("input", std::move(y));
+  SetInputParam("labels", std::move(labels));
+
+  mlpackMain();
+
+  // Check that final output has expected number of rows and colums.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("output").n_rows, 2);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("output").n_cols, 2);
+}
+
+/**
+* Ensure that when we pass optimizer of type lbfgs, we also get the desired
+* shape of output.
+*/
+BOOST_AUTO_TEST_CASE(NCALBFGSTest)
+{
+  arma::mat x;
+  x.randu(3, 100);
+  arma::Row<size_t> labels;
+  labels.zeros(3);
+
+  // Input random data points.
+  SetInputParam("input", std::move(x));
+  SetInputParam("labels", std::move(labels));
+  SetInputParam("optimizer",  std::string("lbfgs"));
+
+  mlpackMain();
+
+  // Check that final output has expected number of rows and colums.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("output").n_rows, 3);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("output").n_cols, 3);
+}
+
+/**
+* Ensure that setting normalize as true results in a
+* different output matrix then when set to false.
+*/
+BOOST_AUTO_TEST_CASE(NCANormalizationTest)
+{
+  // Simple dataset with 6 points and two classes.
+  arma::mat x              = "-0.1 -0.1 -0.1  0.1  0.1  0.1;"
+                             " 1.0  0.0 -1.0  1.0  0.0 -1.0 ";
+  arma::Row<size_t> labels = " 0    0    0    1    1    1   ";
+
+  // Set parameters and set normalize to true.
+  SetInputParam("input", std::move(x));
+  SetInputParam("labels", std::move(labels));
+  SetInputParam("normalize", true);
```

Comment:
So, there is a little bit of trickiness here.  NCA by default uses the SGD optimizer, and the SGD optimizer depends on the random seed since it chooses batches randomly.  In this case, we want to ensure that normalization makes a difference, but as the code currently is, it's likely to give different results simply by virtue of the fact that the order in which batches are visited for SGD will be different.

I think a workaround we can use is for this type of test, we can set `linear_scan` to `true`, so that we will always visit points in the same order.  (The `linear_scan` test is fine as-is, since there all we want to know is that we get a different result there.  In fact, it might be useful to add another test where we run the same data twice with `linear_scan` set, and we should expect the same result.)